### PR TITLE
[dynamo] Preserve identity of mutable attrs on classes built while tracing

### PR DESCRIPTION
**Pull request by aperson30**
## Summary

With `enable_trace_load_build_class=True`, mutating a mutable attribute of a class defined inside the compiled region was silently dropped:

```python
def fn(x):
    class Holder:
        items = []
    Holder.items.append(1)
    return len(Holder.items)   # eager: 1, compiled: 0
```

Such a class is created by `__build_class__` during tracing and has no source, so `resolve_cls_plain_attr` fell through to a sourceless `VariableTracker.build` on every read. Each read therefore produced a distinct VT tracking its own copy of the contents, and a mutation applied to one was invisible to the next read. In CPython `cls.attr` returns the same object every time, so the two diverged with no error.

Classes defined outside the region are unaffected: they have a source (or, when reached through an instance, a `cls_source`) to key the VT on, which is why this only shows up for the traced-class path.

The fix memoizes the VT for mutable class attributes by the identity of the underlying object, in a `SideEffects`-owned map so it participates in clone/restore on speculation restarts and pins the object via `keepalive`. It is applied at both read paths: `UserDefinedClassVariable.resolve_cls_plain_attr` (`Holder.items`) and `UserDefinedObjectVariable.resolve_type_attr` (`instance.items`), which is a separate path and was missed by the first one.

Restricted to `list`/`dict`/`set`: only mutable containers can observe the lost mutation, and keying immutable values by `id()` would be wrong for interned objects such as small ints.

Setting an attribute on such a class (`Holder.v = 3`, which is what blocks `@dataclass` on a locally-defined class) still graph breaks in `BuiltinVariable._call_setattr`. That is a loud failure rather than a wrong result, so it is left for a follow-up.

Fixes #151251

## Test plan

Added `test___build_class___mutable_attr_mutation` to `test/dynamo/test_misc.py`, covering mutation of a list and a dict class attribute read back through both the class and an instance. Verified it fails before the change and passes after:

```
python -m pytest test/dynamo/test_misc.py -k build_class -v
# before: 1 failed, 5 passed, 1 xfailed  (AssertionError on the new test)
# after:  6 passed, 1 xfailed
```

Full files, compared against a baseline run with the change reverted:

```
python -m pytest test/dynamo/test_misc.py -q
# baseline: 6 failed, 796 passed, 21 skipped, 4 xfailed
# patched:  5 failed, 797 passed, 21 skipped, 4 xfailed
# the only difference is the new test; the other 5 failures are pre-existing
# in this environment (inductor needs a C++ compiler this machine lacks)

python -m pytest test/dynamo/test_functions.py -q       # 519 passed
python -m pytest test/dynamo/test_generator.py -q       # 123 passed
python -m pytest test/dynamo/test_mapping_ops.py -q     # 51 passed
python -m pytest test/dynamo/test_error_messages.py -q  # 71 passed
```

`test_generator.py` and `test_mapping_ops.py` both enable `enable_trace_load_build_class` for their whole class, so they exercise this path directly.

`lintrunner` could not run here (its CLANGFORMAT init does not support Windows), so ruff was run directly:

```
ruff check torch/_dynamo/side_effects.py torch/_dynamo/variables/user_defined.py
ruff format --check torch/_dynamo/side_effects.py torch/_dynamo/variables/user_defined.py
```

`check` passes; `format` flags only a pre-existing lambda at `side_effects.py:1697` that reproduces with this change stashed and is far from these edits.

Marking as draft: my environment has no CUDA and no C++ compiler, so the tests above were run against a CPU build with `backend="eager"` and I could not run the GPU/inductor suites. Would appreciate a full CI run before merge.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

## Update: cross-scope HOP mutation now graph-breaks instead of silently dropping

The class-attribute memo is created at whichever scope first reads it. If
that's the top-level scope and a later mutation happens inside a
higher-order-op body (e.g. a `tensor.register_hook` callback, `torch.cond`,
`torch.utils.checkpoint`), the mutation is now caught by the ordinary
cross-scope HOP side-effect check (`is_side_effect_safe`) and graph-breaks,
instead of being silently dropped as it would without this fix. This is a
user-visible behavior change: code that previously compiled (with the
mutation quietly lost) may now graph break instead. Loud over silent is the
intended tradeoff; see
`test___build_class___mutable_attr_mutation_across_hop_scope_graph_breaks`.

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by aperson30** [comment]
@jansel Both blocking findings were correct, confirmed by reproducing the crash locally (KeyError at `side_effects.py:1008`, exactly as described) before fixing.

Fixed in `ce407578`:
1. Collapsed the two methods into `build_sourceless_cls_attr`, which pins the object via a `(item, VariableTracker)` tuple instead of `keepalive`, restoring the `keepalive ⊆ id_to_variable` invariant.
2. Added the missing `sourceless_cls_attrs.clear()`.

Also addressed #3/#4 (the collapse fixes both) and #7 (added `test___build_class___mutable_attr_mutation_with_hop_in_frame`, verified it fails pre-fix and passes now).

#5 and #6 are real but pre-existing, not fixed here.

**Comment by jansel** [comment]
@claude review these changes again

**Comment by aperson30** [comment]
@jansel Addressed the non-blocking notes in 1d2b591:

1/5: Reworded both "no source" comments, since it's not only classes built in-region (also anything reached sourcelessly, e.g. via SourcelessBuilder). Also documented the nested-container gap (a tuple of lists isn't caught by the isinstance check).

4: Added a note next to the existing keepalive NB explaining why sourceless_cls_attrs is excluded from __eq__/diff too.

2: Confirmed your read was right, reproduced the repro locally: reading the memo outside a HOP then mutating inside it now graph breaks (HOP: Unsafe side effect) instead of silently dropping the mutation. Added a test for it.

3: Added fullgraph=True + a frame_count assertion to the existing HOP test so it can't silently degrade into a graph break.

6 (release notes label): already present.

File: torch/_dynamo/side_effects.py [hunk 1]
```diff
@@ -756,6 +768,17 @@ def track_object_existing(
             mutation_type_cls=AttributeMutationExisting,
         )
 
+    def build_sourceless_cls_attr(
+        self, tx: "InstructionTranslatorBase", item: Any
+    ) -> VariableTracker:
+        """Get-or-build the memoized VT for a mutable sourceless class
+        attribute (see sourceless_cls_attrs). Keyed by identity of item."""
+        entry = self.sourceless_cls_attrs.get(id(item))
```
Comment (jansel) [inline torch/_dynamo/side_effects.py:806]: Are we guarding on the aliasing relationships here not changing?  How do we know that we can rely on the ID-matching relationships staying the same between compile time and runtime?

Why isn't there a source for these in the first place?  Maybe the root cause fix here would be to populate the source properly?

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by aperson30** [comment]
@jansel Both checked out, in `dc9d8858`:

1) Aliasing safety: confirmed safe. The tuple pinned `item` for the entry's lifetime, `RestartAnalysis` always builds a fresh `SideEffects`, and the memo never reached codegen.

2) You were right this deserved a real fix. Redesigned `build_sourceless_cls_attr` to route through `id_to_variable` + `AttributeMutationNew`, the same identity system Dynamo already uses for other sourceless objects (Enum/DispatchKey members in `SourcelessBuilder`), instead of the bespoke map. Deleted `sourceless_cls_attrs` entirely. Needed one addition: `codegen_save_tempvars` had no reconstruction path for container VTs under `AttributeMutationNew` (its fallback assumes `UserDefinedObjectVariable`-shaped state), so added one modeled on the existing `ExceptionVariable` case. Full sweep still 828/828, lint clean.

While stress-testing I found a separate, pre-existing bug (also reproduces on the prior commit, not a regression): a mutation before a graph break is lost if the attribute is mutated again after it. Traced it to the resume function's reconstruction not seeing the real object's already-replayed state -- deeper than this PR can fix safely. Filed as #192872.

**Comment by jansel** [comment]
@claude review this

**Comment by aperson30** [comment]
@jansel Confirmed, good catch. Reverted in f0010d063 back to the bespoke sourceless_cls_attrs map (1d2b591) -- it never touches id_to_variable, so it can't shadow a sourced read, and findings 1/3 (codegen non-idempotence, OrderedDict coverage) go away with it. Verified your G=[] and escaping-container repros directly; added a test for the escaping case.

Note: this fixes the regression, not the pre-existing aliasing hazard from note #6 -- still deferred, as before. Agree the real fix is codegen-based, but that's bigger than this PR -- happy to file it separately rather than attempt a third rewrite mid-review.

Full sweep: 830 passed, lint clean.

**Comment by jansel** [comment]
@claude review again

**Comment by aperson30** [comment]
@jansel Addressed 1, 2, and 4 (all optional, all landed in 76f592fa). Finding 1: build_sourceless_cls_attr now checks id_to_variable first (read-only) and reuses a sourced VT if one already exists -- verified your G=[] repro both orderings, sourced-first now replays onto the real object, sourceless-first is unchanged as expected. Finding 2: comment updated. Finding 4: escaping-container test now asserts the literal, twice. Added a PR description note for 3 (the HOP loud-over-silent tradeoff).

831 passed, lint clean.

**Comment by jansel** [comment]
@aperson30 do you have access to claude directly?  I'm seeing a ton of back-and-forth on claude reviews on your PRs.

**Comment by aperson30** [comment]
@jansel Yes, I have Claude I'll start running reviews locally before pushing so the thread doesn't get cluttered with the back-and-forth
